README.md
# RISC-V
A simple RISC-V instruction set architecture simulator written in Java. The program supports reading in programs as binary files and it produces a small text file indicating the content of each of the registers in the register file. The simulator is not pipelined - it implements a simple single-cycle processor.

Guide
=====
The simulator runs a given program when it is run from the command line. The simulator can be set up using a set of constants stored in the top of the IsaSim.java file. These constants are
- ```FILEPATH``` (string) indicating the relative path to the binary file to execute (no default value is provided)
- ```INITIAL_PC``` (integer) indicating the initial value of the program counter (default is ```0```, as programs are read into memory starting from memory address 0)
- ```INITIAL_SP``` (integer) indicating the initial value of the stack pointer (default is ```2^31 - 1 = Integer.MAX_VALUE```)
- ```DEBUGGING``` (boolean) indicating whether debugging prints to the terminal are wanted or not (default is ```true```)

When a program has been executed, the program prints a .txt file and a binary file in the directory in which the test program is stored. The files are named after the program executed followed by ```_reg```. The .txt file contains both the post execution register content from the simulator as well as the expected/correct register content provided to the program in corresponding binary .res files. The binary file contains only the post execution register content.

# 8051

BETA TEST
=========
J51 a Java Intel 8051 family emulator.

    run : java -jar j51.jar

A j51.conf must exist with the supported cpu class names.

FILES on binary distribution
============================
j51.jar	   |	J51 Emulator
---------- | ---
j51.conf   |	Default configuration
diseqc.hex |	Software for diseqc emulator
blink.hex  |	RT Example program require standard P8051
mcb900.hex |	Example from KEIL MCB900 IDE
cmon51.hex |	CMON51 compiled with shared memory
tetris52.hex |	Tetris for CMON51 or P8051 emulator (run at 0x8000)
lpc936.hex |	RT Example for LPC936

Limitations
===========
The J51 disassembler do not work with self modification code like CMON51.

File extension
==============
.flash	| Is for flash memory
------- | ---
.eeprom	| Is for eeprom memory
.misc	| Is for special function memory


CHANGELOG
v1.24 XW 19-06-2022
fixed: errors in JVM

v1.23 XW 04-05-2022
added: nfs

v1.22 XW 04-02-2021
reorded: source files

v1.21 XW 12-06-2020
added: a working filesystem

v1.20 XW 04-04-2020
fixed: read more than one sector (issue #7)
fixed: read more than one cluster (issue #7)

v1.19 XW 14-11-2019
removed: http server

v1.18 XW 12-11-2019
added: shell command (issue #19)
fixed: shell (issue #19)
added: unittest
improved: code (issue #20)

v1.17 XW 11-11-2019
fixed: caches (issue #15)

v1.16 XW 01-09-2019
added: fat long name (issue #16)

v1.15 XW 26-08-2019
improved: filesystem

v1.14 XW 25-08-2019
fixed: time out with 5 times of tries (issue #6)
added: fat32 (issue #8)

v1.13 XW 24-08-2019
fixed: filesystem

v1.12 XW 13-08-2019
fixed: webserver

v1.11 XW 11-08-2019
fixed: dhcp (issue #6)

v1.10 XW 09-08-2019
fixed: receive udp via java api

v1.9 XW 07-08-2019
fixed: send udp via java api

v1.8 XW 05-08-2019
fixed: fat12 (issue #3)

v1.7 XW 04-08-2019
added: keyboard

v1.6 XW 03-08-2019
fixed: ping (issue #2)
added: console

v1.5 XW 01-08-2019
fixed: read from a file
updated: readme
added: network receive (issue #1)
improved: code

v1.4 XW 31-07-2019
added: fat filesystem

v1.3 XW 30-07-2019
added: network sending is working

v1.2 XW 28-07-2019
added: lance driver

v1.1 XW 27-07-2019
added: bioide and BioRAM

v1.0 XW 26-07-2019
intial version
added: AI


src/j51/JInfo.java
/**
 * $Id: JInfo.java 72 2010-07-02 06:56:21Z mviara $
 *
 * 8051 Simulator Information panel.
 *
 * @author Mario Viara
 * @version 1.01
 *
 */
package j51;

import java.awt.*;
import javax.swing.*;
import javax.swing.event.*;
import java.text.*;
import java.beans.*;
   
import j51.intel.*;
import j51.swing.*;

class Block extends JComponent
{
	private Color color;
	private final Dimension size;
	private int perc;
	
	Block(Color color,int w,int h)
	{
		this.color = color;
		size = new Dimension(w,h);
		setPreferredSize(size);
	}

        @Override
	public void paintComponent(Graphics g)
	{
		int h = (size.height * perc ) / 100;
		if (h <= 0)
			h = 1;
		if (h > size.height)
			h = size.height;
		g.setColor(color);
		
		g.fillRect(0,size.height-h,size.width,h);
	}

	public void setColor(Color color)
	{
		this.color = color;
		repaint();
	}

	public void setUsage(int perc)
	{
		this.perc = perc;
		repaint();
	}
}

class CpuUsage extends JPanel
{
	JFixedField perc = new JFixedField(4,true);
	Block	  block;
	JLabel	  label;

	CpuUsage(String name,Color c)
	{
		super(new GridBagLayout());
		GridBagConstraints g = new GridBagConstraints();
		g.gridx = 0;g.gridy = 0;g.gridwidth = 1;g.gridheight = 1;
		g.fill  = GridBagConstraints.BOTH;g.insets = new Insets(2,2,2,2);
		g.anchor = GridBagConstraints.CENTER;
		Dimension dim = perc.getPreferredSize();
		block = new Block(c,dim.width,dim.height*3);
		label = new JLabel(name);
		perc.setEditable(false);
		perc.setHorizontalAlignment(JFixedField.RIGHT);

		add(label,g);g.gridy++;
		add(block,g);g.gridy++;
		add(perc,g);
	}

	public void setUsage(int usage)
	{
		perc.setText(usage+"%");
		block.setUsage(usage);
	}
}

public class JInfo extends J51Panel implements MCS51Performance,ResetListener
{
	JFormattedTextField oscillator = new JFormattedTextField(new DecimalFormat("###,###,##0"));
	JFormattedTextField clock = new JFormattedTextField(new DecimalFormat("###,###,##0"));
	SpinnerNumberModel cycleModel = new SpinnerNumberModel(1,1,16,1);
	JSpinner	cycle = new JSpinner(cycleModel);
	JHexWord	reset;
	JTextField	elapsed;
	
	CpuUsage  min,max,avg,cur;
	private int minCpuUsage,maxCpuUsage,avgCpuUsage;

	public JInfo()
	{
		super("Information");
		GridBagConstraints g = new GridBagConstraints();

		//oscillator	= new JNumField(15,true);
		oscillator.setColumns(15);
		clock.setColumns(15);
		reset		= new JHexWord(true);
		elapsed		= new JTextField(30);
		min		= new CpuUsage("Min",Color.yellow);
		max		= new CpuUsage("Max",Color.red);
		avg		= new CpuUsage("Avg",Color.green);
		cur		= new CpuUsage("Cur",Color.blue);

		min.setUsage(0);
		max.setUsage(100);
		avg.setUsage(50);
		cur.setUsage(25);
		
		JFactory.setBox(oscillator);
		JFactory.setBox(clock);
		JFactory.setBox(cycle);

		JFactory.setBox(reset);

		oscillator.addPropertyChangeListener("value", (PropertyChangeEvent e) -> {
                    Object o = oscillator.getValue();
                    int value = 0;
                    if (o instanceof Number)
                        value = ((Number)o).intValue();
                    //int value = (Integer)oscillator.getValue();
                    cpu.setOscillator(value);
                });
		
		cycle.addChangeListener((ChangeEvent e) -> {
                    cpu.machineCycle((int)(Integer)cycle.getValue());
                });

		oscillator.setToolTipText("Selected oscillator in Hz");
		cycle.setToolTipText("Machina cycle in oscillator clock");
		clock.setToolTipText("Current clock counter");
		clock.setEditable(false);

		oscillator.setHorizontalAlignment(JFixedField.RIGHT);
		clock.setHorizontalAlignment(JFixedField.RIGHT);

		g.gridx = 0;g.gridy = 0;g.gridwidth = 1;g.gridheight = 1;
		g.fill  = GridBagConstraints.NONE;g.insets = new Insets(2,2,2,2);
		g.gridheight = 3;
		
		add(cur,g);g.gridx++;
		add(min,g);g.gridx++;
		add(max,g);g.gridx++;
		add(avg,g);g.gridx++;
		
		g.gridx = 4; g.gridwidth = 1;g.gridheight = 1;
		g.anchor = GridBagConstraints.WEST;
		add(new JLabel("Oscillator"),g);
		g.gridx++;g.anchor = GridBagConstraints.EAST;
		add(oscillator,g);
		g.gridx++;
		g.anchor = GridBagConstraints.WEST;
		add(new JLabel("Cycle"),g);
		g.gridx++;g.anchor = GridBagConstraints.EAST;
		add(cycle,g);
		g.gridx++;g.gridy++;
		g.gridx = 4;
		
		g.anchor = GridBagConstraints.WEST;
		add(new JLabel("Clock"),g);
		g.gridx++;g.anchor = GridBagConstraints.EAST;
		add(clock,g);
		g.gridx++;
		
		add(new JLabel("Reset at"),g);
		g.gridx++;
		add(reset,g);

		g.gridy++;g.gridx = 4; g.gridwidth = 1;g.gridheight = 1;
		g.anchor=GridBagConstraints.SOUTHWEST;
		add(new JLabel("Elapsed"),g);
		g.gridx++;g.gridwidth = 3;
		add(elapsed,g);




	}

        @Override
	public void reset(MCS51 cpu)
	{
		minCpuUsage = 100;
		maxCpuUsage = 0;
		avgCpuUsage = 0;
		updateUsage(0);
	}

	private void updateUsage(int cpu)
	{
		min.setUsage(minCpuUsage);
		max.setUsage(maxCpuUsage);
		avg.setUsage(avgCpuUsage);
		cur.setUsage(cpu);
		
	}
	
        @Override
	public void setCpu(MCS51 cpu)
	{
		super.setCpu(cpu);
		reset(cpu);
		cpu.addPerformanceListener(this);
		cpu.addResetListener(this);
		
	}
	
        @Override
	public void cpuPerformance(int cpu,int elapsed)
	{
		avgCpuUsage = (cpu + avgCpuUsage) / 2;
		if (cpu < minCpuUsage)
			minCpuUsage = cpu;
		if (cpu > maxCpuUsage)
			maxCpuUsage = cpu;
		updateUsage(cpu);
	}

	
        @Override
	public void setEmulation(boolean mode)
	{
			
		oscillator.setEditable(!mode);
		cycle.setEnabled(!mode);
		reset.setEditable(!mode);
	}

	public long updateClock(StringBuffer sb,String name,long v,long d)
	{
		long n = v / d;
		
		if (n > 0)
		{
			sb.append(n).append(" ").append(name).append(" ");
			v -= d * n;
		}

		return v;
		
	}
	
	public void updateClock()
	{
		long elapsed = cpu.clock();
		long tmp;
		
		clock.setValue(elapsed);
		elapsed *= 1000000;
		elapsed /= cpu.getOscillator();
		StringBuffer s = new StringBuffer();
		elapsed = updateClock(s,"day" ,elapsed,1000000L * 60L * 60L * 24L);
		elapsed = updateClock(s,"hour",elapsed,1000000L * 60L * 60L);
		elapsed = updateClock(s,"min", elapsed,1000000L * 60L);
		elapsed = updateClock(s,"sec", elapsed,1000000L);
		elapsed = updateClock(s,"ms",  elapsed,1000L);
		elapsed = updateClock(s,"us",  elapsed,1L);
		
		this.elapsed.setText(s.toString());
			
		
	}

        @Override
	public void update(boolean force)
	{
		oscillator.setValue(cpu.getOscillator());
		updateClock();
		cycle.setValue(cpu.machineCycle());
	}

}
